pkg/storage/ec/client.go

```diff
@@ -254,3 +247,41 @@ func calcPadded(size int64, blockSize int) int64 {
 	}
 	return size + int64(blockSize) - mod
 }
+
+type lazyPieceRanger struct {
+	ranger ranger.RangeCloser
+	dialer dialer
+	node   *proto.Node
+	id     client.PieceID
+	size   int64
+	pba    *pb.PayerBandwidthAllocation
+}
+
+// Size implements Ranger.Size
+func (lr *lazyPieceRanger) Size() int64 {
+	return lr.size
+}
+
+// Size implements Ranger.Close
+func (lr *lazyPieceRanger) Close() error {
+	if lr.ranger == nil {
+		return nil
+	}
+	return lr.ranger.Close()
+}
+
+// Range implements Ranger.Range to be lazily connected
+func (lr *lazyPieceRanger) Range(ctx context.Context, offset, length int64) (io.ReadCloser, error) {
+	if lr.ranger == nil {
+		ps, err := lr.dialer.dial(ctx, lr.node)
+		if err != nil {
+			return nil, err
+		}
+		ranger, err := ps.Get(ctx, lr.id, lr.size, lr.pba)
```

Comment:
hmm, didn't notice this before, but how does `ps` get closed? are we leaking a connection here?

Reply:
I believe they are closed by the DecodeReader here: https://github.com/storj/storj/blob/master/pkg/eestream/decode.go#L210